gradio/event_queue.py

```diff
@@ -0,0 +1,276 @@
+        """
+
+        if live_queue_updates is False and update_intervals == 5:
+            update_intervals = 10
+        cls.SERVER_PATH = server_path
+        cls.LIVE_QUEUE_UPDATES = live_queue_updates
+        cls.MAX_THREAD_COUNT = queue_concurrency_count
+        cls.DATA_GATHERING_STARTS_AT = data_gathering_start
+        cls.UPDATE_INTERVALS = update_intervals
+        cls.DURATION_HISTORY_SIZE = duration_history_size
+        cls.ACTIVE_JOBS = [None] * cls.MAX_THREAD_COUNT
+
+    @classmethod
+    async def init(
+        cls,
+    ) -> None:
+        run_coro_in_background(Queue.notify_clients)
+        run_coro_in_background(Queue.start_processing)
+
+    @classmethod
+    def close(cls):
+        cls.STOP = True
+
+    @classmethod
+    def resume(cls):
+        cls.STOP = False
+
+    # TODO: Remove prints
+    @classmethod
+    async def start_processing(cls) -> None:
+        while not cls.STOP:
+            if not cls.EVENT_QUEUE:
+                await asyncio.sleep(cls.SLEEP_WHEN_FREE)
+                continue
+
+            print(f"Searching for inactive job slots")
+            if not (None in cls.ACTIVE_JOBS):
+                await asyncio.sleep(1)
+                continue
+
+            # Using mutex to avoid editing a list in use
+            async with cls.LOCK:
+                print(f"Start Processing, found inactive job slot, now popping event")
+                event = cls.EVENT_QUEUE.pop(0)
+                print(f"Start Processing, found event, popped event: {event}")
+
+            cls.ACTIVE_JOBS[cls.ACTIVE_JOBS.index(None)] = event
+            run_coro_in_background(cls.process_event, event)
+            run_coro_in_background(cls.gather_data_for_first_ranks)
+            if cls.LIVE_QUEUE_UPDATES:
+                run_coro_in_background(cls.broadcast_estimation)
+
+    @classmethod
+    def push(cls, event: Event):
+        cls.EVENT_QUEUE.append(event)
+
+    @classmethod
+    def clean_job(cls, event: Event):
+        cls.ACTIVE_JOBS[cls.ACTIVE_JOBS.index(event)] = None
+
+    @classmethod
+    async def clean_event(cls, event: Event) -> None:
+        # Using mutex to avoid editing a list in use
+        async with cls.LOCK:
+            # TODO: Might log number of cleaned events in the future
+            try:  # TODO: Might search with event hash to speed up
+                cls.EVENT_QUEUE.remove(event)
+            except ValueError:
+                pass
+
+    @classmethod
+    async def gather_data_for_first_ranks(cls) -> None:
+        """
+        Gather data for the first x events.
+        """
+        await asyncio.gather(
+            *[
+                cls.gather_event_data(event)
+                for event in cls.EVENT_QUEUE[: cls.DATA_GATHERING_STARTS_AT]
+            ]
+        )
+
+    @classmethod
+    async def gather_event_data(cls, event: Event) -> None:
+        """
+        Gather data for the event
+
+        Args:
```

Comment:
Here and elsewhere throughout the code, can you use `Parameters` instead of `Args` to be consistent with docstrings throughout the library

Reply:
Will do, but I actually don't like the current status, different standards in the codebase and I don't know which standard we're following. Will follow up this is in another issue.